src/core/recv_buffer.c

```diff
@@ -977,13 +1124,16 @@ QuicRecvBufferDrain(
     CXPLAT_DBG_ASSERT(FirstRange->Low == 0);
     do {
         BOOLEAN PartialDrain = (uint64_t)RecvBuffer->ReadLength > DrainLength;
-        if (PartialDrain ||
-            (QuicRangeSize(&RecvBuffer->WrittenRanges) > 1 &&
-             RecvBuffer->BaseOffset + RecvBuffer->ReadLength == FirstRange->Count)) {
+        // TODO guhetier: Seems wrong, if the end of the chunk aligns with the end of the first range, we still need a full drain
```

Comment:
For stuff like this, can you update the unittest to verify?

Reply:
That's the plan!

Reply:
The fix was not trivial here - hopefully I got it right, at least, old and new tests are passing.
Multiple variables are being used in subtly different ways depending on the mode (not sure if it is intentional or not), making those few lines much trickier than expected.
In particular, this code works for Single/Circular mode only because `ReadLength` is actually the readable size on the last chunk for these modes, not the first as documented.